code/game/machinery/vendors/vending.dm

```diff
@@ -0,0 +1,880 @@
+		if(user.put_in_hands(inserted_item))
+			put_on_turf = FALSE
+	if(put_on_turf)
+		var/turf/T = get_turf(src)
+		inserted_item.forceMove(T)
+	inserted_item = null
+	SStgui.update_uis(src)
+
+/obj/machinery/economy/vending/emag_act(mob/user)
+	emagged = TRUE
+	to_chat(user, "You short out the product lock on [src]")
+
+/obj/machinery/economy/vending/attack_ai(mob/user)
+	return attack_hand(user)
+
+/obj/machinery/economy/vending/attack_ghost(mob/user)
+	return attack_hand(user)
+
+/obj/machinery/economy/vending/attack_hand(mob/user)
+	if(stat & (BROKEN|NOPOWER))
+		return
+
+	if(seconds_electrified != 0)
+		if(shock(user, 100))
+			return
+
+	ui_interact(user)
+	wires.Interact(user)
+
+/obj/machinery/economy/vending/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		var/estimated_height = 100 + min(length(product_records) * 34, 500)
+		if(length(prices) > 0)
+			estimated_height += 100 // to account for the "current user" interface
+		ui = new(user, src, ui_key, "Vending",  name, 470, estimated_height, master_ui, state)
+		ui.open()
+
+/obj/machinery/economy/vending/ui_data(mob/user)
+	var/list/data = list()
+
+	data["usermoney"] = 0
+	data["inserted_cash"] = cash_transaction
+	data["user"] = null
+	if(ishuman(user))
+		var/obj/item/card/id/C
+		var/mob/living/carbon/human/H = user
+		C = H.get_idcard(TRUE)
+		if(!C && istype(H.wear_pda, /obj/item/pda))
+			var/obj/item/pda/P = H.wear_pda
+			if(istype(P.id, /obj/item/card/id))
+				C = P.id
+		if(istype(C))
+			var/datum/money_account/A = C.get_card_account()
+			if(A)
+				data["user"] = list()
+				data["user"]["name"] = A.account_name
+				data["usermoney"] = A.credit_balance
+				data["user"]["job"] = C.rank ? C.rank : "No Job"
+
+	data["stock"] = list()
+	for (var/datum/data/vending_product/R in product_records + coin_records + hidden_records)
+		data["stock"][R.name] = R.amount
+	data["extended_inventory"] = extended_inventory
+	data["vend_ready"] = vend_ready
+	data["coin_name"] = coin ? coin.name : FALSE
+	data["panel_open"] = panel_open ? TRUE : FALSE
+	data["speaker"] = shut_up ? FALSE : TRUE
+	data["item_slot"] = item_slot // boolean
+	data["inserted_item_name"] = inserted_item ? inserted_item.name : FALSE
+	return data
+
+
+/obj/machinery/economy/vending/ui_static_data(mob/user)
+	var/list/data = list()
+	data["chargesMoney"] = length(prices) > 0 ? TRUE : FALSE
+	data["product_records"] = list()
+	var/i = 1
+	for(var/datum/data/vending_product/R in product_records)
+		var/list/data_pr = list(
+			path = replacetext(replacetext("[R.product_path]", "/obj/item/", ""), "/", "-"),
+			name = R.name,
+			price = (R.product_path in prices) ? prices[R.product_path] : 0,
+			max_amount = R.max_amount,
+			req_coin = FALSE,
+			is_hidden = FALSE,
+			inum = i
+		)
+		data["product_records"] += list(data_pr)
+		i++
```

Comment:
```suggestion
			inum = i++
		)
		data["product_records"] += list(data_pr)
```
Can also be done for the others.
`i++` basically says. Return i, then increase it. Opposed to `++i` which increases i and then returns it